### ISSUE TITLE
Async azure functions will be always succes and does not have extra deliveries.

### ISSUE THREAD

**Issue by RikShoeby** [Azure/azure-functions-dotnet-worker#607]
With our current azure functions (.net core 5) we have had a big production problem. The azure functions stopped running at the 23-08 for days until we looked into the app and it was starting again. I still have to look into that. Lets focus on the first problem I found.

Luckily all messages where still on the "Dead-letter". We put them again on the "Active" list and the messages where processed. Not all the messages where successful due timeouts of a API. For some reason all the messages where handled and gone instead of doing the 5 retries and there after put it on the dead letter.

Some of the RpcExceptions where unhandled but where not shown in the failed list on the left..
![image](https://user-images.githubusercontent.com/89585999/131315260-d5348b37-8604-4290-b6af-3cf3492cd962.png)

After a lot of testing and creating a new small azure function I found something i did not expect..
The function Test.Test1-async and Test.Test1 are the same with only the async keyword as difference.

```
public class Test1Message
    {
        [JsonPropertyName("id")]
        [Required]
        public int? Id { get; set; }
    }

    public class Test1
    {
        [Function("Test.Test1-async")]
        public async void RunAsync([ServiceBusTrigger("test-test1", "shoeby-order-service", Connection = "ServiceBus")] string mySbMsg, FunctionContext context)
        {
            var logger = context.GetLogger("OrderUpdated");
            logger.LogInformation($"C# ServiceBus topic trigger (test-test1) function processed message: {mySbMsg}");

            var messageModel = JsonSerializer.Deserialize<Test1Message>(mySbMsg);

            if (messageModel?.Id == null)
            {
                throw new ArgumentNullException(nameof(messageModel));
            }


            if (messageModel.Id < 10)
            {
                throw new NotSupportedException("Id must be larger then 10!!");
            }
        }

        [Function("Test.Test1")]
        public void Run([ServiceBusTrigger("test-test1", "shoeby-order-service", Connection = "ServiceBus")] string mySbMsg, FunctionContext context)
        {
            var logger = context.GetLogger("OrderUpdated");
            logger.LogInformation($"C# ServiceBus topic trigger (test-test1) function processed message: {mySbMsg}");

            var messageModel = JsonSerializer.Deserialize<Test1Message>(mySbMsg);

            if (messageModel?.Id == null)
            {
                throw new ArgumentNullException(nameof(messageModel));
            }


            if (messageModel.Id < 10)
            {
                throw new NotSupportedException("Id must be larger then 10!!");
            }
        }
    }
```

The results of the Test.Test1
![image](https://user-images.githubusercontent.com/89585999/131316035-15ffcbdd-60df-43c9-80a5-0536bea229a7.png)
Also the message did go to the dead-letter after 5 tries.

The results of the Test.Test1-async:
![image](https://user-images.githubusercontent.com/89585999/131316519-ed5c60ed-265b-47e6-9949-2eb96635496d.png)

**Comment by kshyju** [comment]
Looking at the signature of `Test.Test1-async` function method, I see a red flag. The `async void` !

You should use `async void` only if you are writing an async event handler. For normal use cases like yours, you should be using `async Task` or `async Task<T>`, if your method is doing some asynchronous work inside.

The incorrect usage of `async void` could be the cause of the behavior you are seeing.  For a normal async method (which returns `Task` or `Task<T>`), when an exception is thrown out of that method, that exception is captured and put on the returned Task.  When the calling code await this task, the first exception present in the task is re-thrown.  But, If you use `async void` instead of using `async Task`, there is no Task to put this exception on. The exception will be posted to thread pool and an unhandled exception on thread pool will cause the whole process to terminate.

I do not see any asynchronous work inside your method. In that case. do not use the `async` keyword at all. Remember, when you add the async keyword to a method, the compiler [will generate a state machine for your method](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwKwG5lQDMmM6AwsgN7Lq3oD096ATgKYC2A9gG6uYAc6ANasAngHdOzYOgAundO1ayAFpxkBDAHYyAzqz4BzVltbMNs1jIDGAYnSctLAJaGVsmnSIDMAFnQASgCuWgCCuqJa1gAUWBjsogDKAEYAsrqGADSY2LLOjmSOlgAesujWRaylAJSetNRIdE3o3BrMiqy6uhrGqeqsADboALzopuLoACqdsnCpnd3G0dX4SHXNzgBm6NFKXT2sfcCDAPwAdACSMsOjWkEDA7WNzfXrL7SqzJwT4+ihzIYgkotLIAHL3AYAUWK1lYAAc8o5oloNEpOJtdgsDkdBtUVm86ABfZAE2hbHZ7RaHfoDS4yAA86DgqCe73QDTZTU+3zGrAmoM4skSQThcKklmA0NhCPyWmiACIroogroysk+AM2sZ2qoTEzUABCA3y/HPF7Es3oC3rbywKYzOZY4zrDnvbzOEHoJWUdD6WS4dDGf1W9YWpq20hQXKywogqoeS2u2gWwlAA===).  

Please take a look at [this excellent async usage guidelines document](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md) David Fowler has put together.

In short, use `Task` or `Task<T>` as the return type of your method if your method is doing some asynchronous work.

Hope that helps, let us know if you still have issues after fixing the incorrect async usage on your method. Thanks!

**Comment by RikShoeby** [comment]
Thanks for the response!

You are correct only use the async keyword if there is async work somewhere down the tree. In the real case that will be async call's on the HttpCliënt. For the reproduce scenario I created new functions with the outline of the production functions. And now you point to it, how stupid of us not to see that the async azure function has to return a task (just like all the rest of the code and api's).

With that addition the testcase was solved, but the logging is not working in debug mode (when deployed there is logging in the stream..).
Now I can look into how to not retry the message as fast as possible..

**Comment by RikShoeby** [comment]
With the next code I see retries on failure and logging when deployed. The logging does not work when debugging. 
    ```
public class Test1Message
    {
        [JsonPropertyName("id")]
        [Required]
        public int? Id { get; set; }
    }

    public class Test1
    {
        private readonly IHttpClientFactory _httpClientFactory;

        public Test1(IHttpClientFactory httpClientFactory)
        {
            _httpClientFactory = httpClientFactory;
        }

        [Function("Test.Test1-async-web")]
        public async Task RunAsyncWeb(
            [ServiceBusTrigger("test-test1", "shoeby-order-service", Connection = "ServiceBus")]
            string mySbMsg, FunctionContext context)
        {
            var logger = context.GetLogger("OrderUpdated");
            logger.LogInformation($"C# ServiceBus topic trigger (test-test1) function processed message: {mySbMsg}");

            var messageModel = JsonSerializer.Deserialize<Test1Message>(mySbMsg);

            if (messageModel?.Id == null)
            {
                throw new ArgumentNullException(nameof(messageModel));
            }


            if (messageModel.Id < 10)
            {
                throw new NotSupportedException("Id must be larger then 10!!");
            }

            var request = new HttpRequestMessage(new HttpMethod("GET"),
                $"https://www.google.com/search?q={messageModel.Id}");
            using (var response = await _httpClientFactory.CreateClient().SendAsync(request))
            {
                response.EnsureSuccessStatusCode();

                var content = await response.Content.ReadAsStringAsync();

                logger.LogInformation($"HttpCall response received:{content}");
            }


            logger.LogInformation($"Message: {mySbMsg} processed");
        }
    }
```

**Comment by RikShoeby** [comment]
With the example above I get some strange logging/app insights:
![image](https://user-images.githubusercontent.com/89585999/131565188-f79963ad-e0f2-4efe-8c7b-cffe3e0bf9ee.png)

I see 3 times the same failure as exception, that is a bit anoying.

It gets really a thing if you look into errors. Then you see 5 failed calls because of the 5 retries, then x3 in the RpcException.. Why is it there visible 3x?
![image](https://user-images.githubusercontent.com/89585999/131565497-50dc26c1-f64c-4a1d-babd-35585ad23b9e.png)

**Comment by brettsam** [comment]
I've spawned a separate issue here so we can track an improvement to the `async void` scenario -- https://github.com/Azure/azure-functions-dotnet-worker/issues/610.

**Comment by kshyju** [comment]
@RikShoeby When an unhandled exception happens, it is possible that the functions framework code(host) logs the error in multiple layers, causing more than one error log entry. You should be able to query these exception log entries in application insights for a single invocation and see the different dimensions/meta data present in each of those entries.  You [may control the logging](https://docs.microsoft.com/en-us/azure/azure-functions/functions-host-json#logging) from the function app by adjusting the `logging` settings of the host.json, as needed.
